crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -291,8 +291,15 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
             .chain(comparators)
             .tuple_windows::<(&ruff_python_ast::Expr, &ruff_python_ast::Expr)>();
         let mut constraints = NarrowingConstraints::default();
+
+        let mut last_rhs_ty: Option<Type> = None;
+
         for (op, (left, right)) in std::iter::zip(&**ops, comparator_tuples) {
+            let lhs_ty = last_rhs_ty.unwrap_or_else(|| {
+                inference.expression_ty(left.scoped_expression_id(self.db, scope))
+            });
             let rhs_ty = inference.expression_ty(right.scoped_expression_id(self.db, scope));
+            last_rhs_ty = Some(rhs_ty);
```

Comment:
This looks like it's improving support for chained comparisons? It doesn't appear directly connected to `LiteralString`.

I'm not immediately seeing what test added in this PR requires this support?

Unless this is definitely required in this PR for reasons I'm not seeing, I think it would be easier to evaluate in a separate PR, accompanied by tests which specifically require this?